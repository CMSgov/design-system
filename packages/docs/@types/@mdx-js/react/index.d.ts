declare module '@mdx-js/react' {
  const MDXProvider: React.ComponentType;
  export { MDXProvider };
}
