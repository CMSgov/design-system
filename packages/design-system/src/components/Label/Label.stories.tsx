import React from 'react';
import Label from './Label';
import type { Meta, StoryObj } from '@storybook/react';

const meta: Meta<typeof Label> = {
  title: 'Components/Label',
  component: Label as any,
  args: {
    children: 'Sample Label',
  },
};
export default meta;

type Story = StoryObj<typeof Label>;

export const LabelDefault: Story = {};

export const InverseLabel: Story = {
  args: {
    inversed: true,
  },
  parameters: {
    // Must supply `layout: 'fullscreen'` when we use `onDark: true`
    onDark: true,
    layout: 'fullscreen',
  },
};
