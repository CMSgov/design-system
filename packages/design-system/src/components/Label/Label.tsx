import React from 'react';
import classNames from 'classnames';
import { Hint } from '../Hint';
import { InlineError } from '../InlineError';

export type LabelComponent = 'label' | 'legend';
export interface LabelProps {
  /**
   * Label text or HTML.
   */
  children: React.ReactNode;
  /**
   * Additional classes to be added to the root element.
   */
  className?: string;
  /** The root HTML element used to render the label */
  component?: LabelComponent;
  /**
   * @deprecated Hints are now their own component called `Hint`.
   * @hide-prop [Deprecated]
   *
   * Enable the error state by providing an error message.
   */
  errorMessage?: React.ReactNode;
  /**
   * @deprecated The Label is no longer responsible for rendering the
   * error element from a string. A InlineError should be passed to it which
   * already has an errorId applied.
   * @hide-prop [Deprecated]
   */
  errorId?: string;
  /**
   * The ID of the field this label is for. This is used for the label's `for`
   * attribute and any related ARIA attributes, such as for the error message.
   */
  fieldId?: string;
  /**
   * @deprecated Hints are now their own component called `Hint`.
   * @hide-prop [Deprecated]
   *
   * Additional hint text to display
   */
  hint?: React.ReactNode;
  /**
   * @deprecated Please render your error message directly with the 'InlineError' component instead of passing to `Label`
   * @hide-prop [Deprecated]
   *
   * The ID of the hint element
   */
  hintId?: string;
  /**
   * A unique `id` for the label element. Useful for referencing the label from
   * other components with `aria-describedby`.
   */
  id?: string;
  /**
   * Set to `true` to apply the "inverse" theme
   */
  inversed?: boolean;
  /**
   * Text showing the requirement (ie. "Optional", or "Required").
   * In most cases, this should be used to indicate which fields are optional.
   * See the [form guidelines]({{root}}/guidelines/forms/) for more info.
   */
  requirementLabel?: React.ReactNode;
  /**
   * @deprecated Please just use `className` instead
   * @hide-prop [Deprecated]
   */
  textClassName?: string;
}

type LabelComponentProps = React.ComponentPropsWithRef<'label'> &
  React.ComponentPropsWithRef<'legend'> &
  LabelProps;

/**
 * For information about how and when to use this component,
 * [refer to its full documentation page](https://design.cms.gov/components/form-label/).
 */
export const Label = (props: LabelComponentProps) => {
  const {
    fieldId,
    id,
    children,
    component,
    hint,
    hintId,
    textClassName,
    className,
    inversed,
    errorMessage,
    errorId,
    requirementLabel,
    ...labelProps
  } = props;

  if (process.env.NODE_ENV !== 'production' && textClassName) {
    console.warn(
      "[Deprecated]: Please use the 'className' prop instead of 'textClassName'. This prop is deprecated and will be removed in a future release."
    );
  }

  if (process.env.NODE_ENV !== 'production' && (hint || hintId)) {
    console.warn(
      "[Deprecated]: Hints are now their own component called 'Hint'. Please render your 'Hint' directly instead of passing them to 'Label'."
    );
  }

  if (process.env.NODE_ENV !== 'production' && errorMessage) {
    console.warn(
      "[Deprecated]: Passing 'errorMessage' to the 'Label' component is now deprecated. Please render your error message directly with the 'InlineError' component instead."
    );
  }

  let hintElement;
  if (hint || requirementLabel) {
    hintElement = (
      <Hint requirementLabel={requirementLabel} inversed={inversed} id={hintId}>
        {hint}
      </Hint>
    );
  }

  let errorElement = errorMessage;
  if (typeof errorMessage === 'string') {
    errorElement = <InlineError id={errorId}>{errorMessage}</InlineError>;
  }

  const ComponentType = component;
  const classes = classNames(
    'ds-c-label',
    className,
    inversed && 'ds-c-label--inverse',
    textClassName
  );

  return (
    <>
      <ComponentType className={classes} htmlFor={fieldId} id={id} {...labelProps}>
        {children}
      </ComponentType>
      {hintElement}
      {errorElement}
    </>
  );
};

Label.defaultProps = { component: 'label' };

export default Label;
