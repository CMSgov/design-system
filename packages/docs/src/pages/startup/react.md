---
title: Using React components
weight: 3
---

The examples below assume you've installed the design system using NPM and have already setup your build system.

<h4 class="ds-h4 ds-u-font-size--base">Default imports</h4>

Individual components can be imported from their individual export file.

```jsx
import Button from '@cmsgov/design-system-core/dist/components/Button/Button';
```

<h4 class="ds-h4 ds-u-font-size--base">Named imports</h4>

Components can also be imported using the shorter syntax below.

<div class="ds-c-alert ds-c-alert--warn ds-u-margin-bottom--2 ds-u-font-size--small">
  <div class="ds-c-alert__body">
    <h3 class="ds-c-alert__heading ds-u-font-size--base">Performance note</h3>
    <p class="ds-c-alert__text">
      This approach may result in a much larger file than you intend. Depending on what module bundler you use, all of the design system's React components may be included in the bundled file even if you didn't specifically import them. This can be avoided by enabling features like [tree shaking in Webpack](https://webpack.js.org/guides/tree-shaking/).
    </p>
  </div>
</div>

```jsx
import { Button } from '@cmsgov/design-system-core';
```

[View an example](https://github.com/CMSgov/design-system/blob/master/examples/react-app/src/scripts/index.js)

<h2 id="examples" class="ds-h2 ds-u-color--primary-darker">Examples</h2>

Additional examples of the design system in use can be viewed on GitHub. These projects demonstrate the various ways you can incorporate the design system into your development process and various use cases.

<a href="https://github.com/CMSgov/design-system/tree/master/examples/" class="ds-c-button">Browse example projects</a>

<h2 id="need-help" class="ds-h2 ds-u-color--primary-darker">Need help or ran into an issue?</h2>

If you're having trouble installing or setting up the design system, or if you think you've found a bug, feel free to [open an issue on GitHub](https://github.com/CMSgov/design-system/issues).
