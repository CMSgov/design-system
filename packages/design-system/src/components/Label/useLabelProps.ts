import type * as React from 'react';
import { LabelProps } from '../Label';

export interface UseLabelPropsProps {
  /**
   * Label for the field.
   */
  label: React.ReactNode;
  /**
   * Additional classes to be added to the label
   */
  labelClassName?: string;
  /**
   * A unique `id` to be used on the field label. If one isn't provided, a unique ID
   * will be generated.
   */
  labelId?: string;
  /**
   * A unique `id` for the field element. Useful for referencing the field from
   * other components with `aria-describedby`.
   */
  id?: string;
  /**
   * Set to `true` to apply the "inverse" color scheme
   */
  inversed?: boolean;
  /**
   * Determines whether the label should be hidden from screen readers.
   * Set to `true` to apply `aria-hidden="true"` to the label.
   * This is useful in cases where the label does not provide valuable context for screen reader users,
   * such as when the associated input already has an accessible name.
   */
  labelHidden?: boolean;
}

/**
 *
 * Note that this does not return the `fieldId` in the props. That is up to the component
 * to apply to the `Label` itself. Some components purposely do not set a `for` attribute
 * on their labels. These are typically components with multiple inputs wrapped in a
 * `fieldset` like `ChoiceList`, `MultiInputDateField`, and `MonthPicker`.
 */
export function useLabelProps<T extends UseLabelPropsProps>(props: T): Omit<LabelProps, 'fieldId'> {
  const labelId = props.labelId ?? `${props.id}__label`;

  const { label, labelClassName, inversed, labelHidden } = props;

  return {
    children: label,
    className: labelClassName,
    id: labelId,
    inversed,
    labelHidden,
  };
}

export default useLabelProps;
