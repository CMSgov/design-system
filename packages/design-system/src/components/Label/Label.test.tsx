import React from 'react';
import Label from './Label';
import { render, screen } from '@testing-library/react';

describe('Label', () => {
  const labelText = 'Hello world';

  it('renders label text', () => {
    const props = { fieldId: 'name' };
    render(<Label {...props}>{labelText}</Label>);

    const label = screen.getByText(labelText);
    expect(label).toBeInTheDocument();
    expect(label).toMatchSnapshot();
  });

  it('renders error messages', () => {
    const props = {
      errorMessage: <span>Nah, try again.</span>,
      fieldId: 'name',
    };
    render(<Label {...props}>{labelText}</Label>);

    const error = screen.getByText('Nah, try again.');
    expect(error).toBeInTheDocument();
  });

  it('renders hint string', () => {
    const props = { hint: 'President #44' };
    render(<Label {...props}>{labelText}</Label>);

    const hint = screen.getByText('President #44');
    expect(hint).toBeInTheDocument();
  });

  it('renders hint node', () => {
    const props = {
      hint: <strong>President #44</strong>,
    };
    render(<Label {...props}>{labelText}</Label>);

    const hint = screen.getByText('President #44');
    expect(hint).toBeInTheDocument();
  });

  it('renders requirementLabel', () => {
    const props = {
      requirementLabel: <em>It is really optional</em>,
    };
    const { container } = render(<Label {...props}>{labelText}</Label>);

    const requirement = container.querySelector('label');
    expect(requirement).toBeInTheDocument();
    expect(requirement).toMatchSnapshot();
  });

  it('adds punctuation to requirementLabel when hint is also present', () => {
    interface PropDef {
      hint: React.ReactNode;
      requirementLabel: React.ReactNode;
    }
    const props: PropDef = { hint: 'Hint', requirementLabel: 'Optional' };
    const { container } = render(<Label {...props}>{labelText}</Label>);

    const label = container.querySelector('label');
    expect(label).toMatchSnapshot();
  });

  it('renders as a legend element', () => {
    const { container } = render(<Label component="legend">{labelText}</Label>);

    const legend = container.querySelector('legend');
    expect(legend).toBeInTheDocument();
    expect(legend).toMatchSnapshot();
  });

  it('is inversed', () => {
    const props = {
      hint: 'Foo',
      inversed: true,
    };
    render(<Label {...props}>{labelText}</Label>);

    const inversedHint = screen.getByText('Foo');
    expect(inversedHint).toHaveClass('ds-c-hint ds-c-hint--inverse');
    expect(inversedHint).toMatchSnapshot();
  });

  it('supports additional attributes', () => {
    const props = { className: 'ds-u-foo', 'aria-label': 'testing aria' };
    const { container } = render(<Label {...props}>{labelText}</Label>);

    const label = container.querySelector('label');
    expect(label).toHaveAttribute('aria-label', 'testing aria');
    expect(label).toMatchSnapshot();
  });
});
