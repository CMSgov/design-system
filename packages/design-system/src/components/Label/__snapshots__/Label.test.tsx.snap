// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Label adds punctuation to requirementLabel when hint is also present 1`] = `
<label
  class="ds-c-label"
>
  Hello world
</label>
`;

exports[`Label is inversed 1`] = `
<div
  class="ds-c-hint ds-c-hint--inverse"
>
  Foo
</div>
`;

exports[`Label renders as a legend element 1`] = `
<legend
  class="ds-c-label"
>
  Hello world
</legend>
`;

exports[`Label renders label text 1`] = `
<label
  class="ds-c-label"
  for="name"
>
  Hello world
</label>
`;

exports[`Label renders requirementLabel 1`] = `
<label
  class="ds-c-label"
>
  Hello world
</label>
`;

exports[`Label supports additional attributes 1`] = `
<label
  aria-label="testing aria"
  class="ds-c-label ds-u-foo"
>
  Hello world
</label>
`;
